### ISSUE TITLE
[Feature Request] Make apparent in fully-expanded code the identifer that `contract-out` uses for wrapper definitions

### ISSUE THREAD

**Issue by greghendershott** [racket/racket#3733]
*TL;DR* `contract-out` should make apparent in fully-expanded code the identifier is uses for the wrapper it defines.

# Context

I want to open an issue or PR for `drracket-tool-lib` that proposes having `check-syntax` call `syncheck:add-definition-target` for the wrapper functions defined by `contract-out` forms.

As a result, a tool could use `drracket/check-syntax` to find the **location** (as in `syntax-position` and `syntax-span`) of the wrapper definitions in `contract-out` forms.

Being able to find the **location** of definitions is important to support a variety of use cases:

- DrRacket the IDE could have a full "Go To Definition" command (not just "Open Defining File").
- So-called "refactoring" tools like a reliable **multi-file** "Rename Definitions and Uses" command.

The fly in the ointment is that `racket/contract` currently makes it impossible (AFAICT) to find the definition of a `contract-out` wrapper in fully-expanded code -- at least in a principled, simple, or reliable way.

As a result, any PR I submit to `drracket-tool-lib` is going to be justifiably criticized for using hacks. So I am proactively opening this issue, first. :smile:

Seriously, I think this falls into two distinct buckets: A change to drracket, which requires an enabling change to `contract-out`.

# Background / Status quo

Given some function `foo`, `contract-out` defines a wrapper function named `provide/contract-id-foo.1`, and, exports that renamed as `foo`.

As a result, `provide/contract-id-foo.1` is the source-identifier reported by `identifier-binding` (and therefore `syncheck:add-jump-to-definition`).

The fact that it's a contract wrapper function is immaterial to `identifier-binding`; the exported thing is the wrapper, and the defined name of the wrapper is `provide/contract-id-foo.1` --- not `foo`.

As a result, if a tool asks `identifier-binding`, in a namespace where "foo" was imported, "Where is `foo` defined?", it will answer, "Go look in this module path for `provide/contract-id-foo.1`. And if a tool wants to know the source location of that definition, it must examine the fully-expanded syntax looking for something like `(define-values (provide/contract-id-foo.1) ___)`.

Or, let's say that a tool proactively analyzes a file to record the locations of  definitions -- storing them under the name that `identifier-binding` will report as the source-id.

Today, AFAICT, this can only be done with a hack that is probably fragile. Something like:

```racket
(define (analyze-provide/contract-transformers path stx)

  (define (symbolic-compare? x y)
    (eq? (syntax-e x) (syntax-e y)))

  (define (handle-file-module stx)
    (syntax-case* stx (module) symbolic-compare?
      [(module _id _lang (_mb . es))
       (handle-module-level '() #'es)]))

  (define (handle-module-level submods es)
    (for ([e (in-syntax es)])
      (syntax-case* e
          (#%app
           define-syntaxes
           make-provide/contract-transformer
           make-provide/contract-arrow-transformer
           quote-syntax
           module
           module*)
          symbolic-compare?
        [(define-syntaxes
           (_)
           (#%app
            make-provide/contract-transformer
            (quote-syntax wrapper)
            (quote-syntax _)
            (quote-syntax original)
            . _))
         (add-def-for-contract path submods #'original #'wrapper)]
        [(define-syntaxes
           (_)
           (#%app
            make-provide/contract-arrow-transformer
            (quote-syntax wrapper)
            (quote-syntax _)
            (quote-syntax original)
            . _))
         (add-def-for-contract path submods #'original #'wrapper)]
        [(module id _lang . es)
         (handle-module-level (cons (syntax-e #'id) submods) #'es)]
        [(module* id _lang . es)
         (handle-module-level (cons (syntax-e #'id) submods) #'es)]
        [_ (void)])))

  (define (add-def-for-contract path submods original wrapper)
    (define beg (syntax-position original))
    (define span (syntax-span original))
    (cond
      [(and beg span)
       ;; FIXME: I don't yet understand where the ".1" suffix comes from;
       ;; fragile?!?
       (define wrapper-sym (string->symbol (~a (syntax-e wrapper) ".1")))
       (add-def path beg (+ beg span) submods wrapper-sym)]
      [else
       (log-warning "no syntax location for ~v in ~v" (syntax-e original) path)]))

  (handle-file-module stx))
```

Where `add-def` records this information to a database. Or, if the code were part of `drracket/check-syntax`, would result in a call to `syncheck:add-definition-target`.

**Comment by rfindler** [comment]
I am seeing slightly different behavior in DrRacket that I cannot quite reconcile with the description above so I'm asking for help! 

With this in tmp.rkt:
```
#lang racket
(require "tmp2.rkt")
exported
```

and this in tmp2.rkt:
```
#lang racket
(require "tmp3.rkt")
(provide
 (contract-out
  [exported (-> integer? integer?)]))
```
and this in tmp3.rkt:
```
#lang racket
(define (exported x) x)
(provide exported)
```

then I click go from the use of `exported` in tmp.rkt to the file tmp2.rkt and then from the use of `exported` inside the `contract-out` to the definition in tmp3.rkt.

Does this not work for you or am I missing something about how it should be working differently or do I need a different example? Sorry-- just trying to understand what exactly doesn't work.

**Comment by greghendershott** [comment]
In other words unless I'm being dense, I'm talking about something you can't do today in Dr Racket.

The motivation is this part above:

> 
> As a result, a tool could use `drracket/check-syntax` to find the **location** (as in `syntax-position` and `syntax-span`) of the wrapper definitions in `contract-out` forms.
> 
> Being able to find the **location** of definitions is important to support a variety of use cases:
> 
> - DrRacket the IDE could have a full "Go To Definition" command (not just "Open Defining File").
> - So-called "refactoring" tools like a reliable **multi-file** "Rename Definitions and Uses" command.

**Comment by rfindler** [comment]
re "click go": I meant to right click on the identifier and go to the definition, using the menu item that variously says "Open Defining File" or "Jump to Definition (in other file)".

... oh but now I'm starting to see what I was missing -- Check Syntax *never* says "Jump to Definition (in other file)" for the occurrence in tmp.rkt (whereas for the one in tmp2.rkt, it will say that sometimes, depending on if it has analyzed tmp3.rkt or not).  Is that a symptom of the issue you're seeking to fix?

**Comment by greghendershott** [comment]
I only ever see "Open Defining File".

I never see a menu option "Jump to Definition (in other file)".  I tried with both manual vs. automatic background expansion. I think I tried opening the files in various orders. So I'm not sure what's going wrong, for me.

In fact although I'm not a heavy Dr Racket user I don't think I've _ever_ seen "Jump to Definition (in other file)". I didn't realize it had such a command! When was this added?

Now I'm curious to go through the Dr Racket source and try to find where this is, and how it works. And how it even _could_ work for contract wrappers -- because at the moment I don't see how, unless `syncheck:add-definition-target` is being called for the wrappers like `make-provide/contract-foo.1`, which I'm not seeing in my own use of it.

**Comment by greghendershott** [comment]
I got a bit side-tracked there with other things that confused me, but, you wrote:

> ... oh but now I'm starting to see what I was missing -- Check Syntax never says "Jump to Definition (in other file)" for the occurrence in tmp.rkt (whereas for the one in tmp2.rkt, it will say that sometimes, depending on if it has analyzed tmp3.rkt or not). Is that a symptom of the issue you're seeking to fix?

Yes!

**Comment by samth** [comment]
You only see that option if the other file is already open in DrRacket, and
background expansion has already completed for the other file.

**Comment by greghendershott** [comment]
@samth: I thought I'd tried that permutation but I'll double-check. Thanks.

@rfindler: 

In my use of `drracket/check-syntax`, what I see:

I _do_ already get either `syncheck:add-jump-to-definition` or `syncheck:add-arrow-xxx` for the _use_ of a contract wrapper. It's already doing that "half" of it.  It knows the thing exists, and for `syncheck:add-jump-to-definition` it supplies the wrapper name like `make-provide/contract-foo.1` (re Slack convo, these names are already exposed).

But it's not doing the "other half": a `syncheck:add-definition-target` for the _definition_ of the wrapper within the `contract-out` form.

**Comment by greghendershott** [comment]
To offer a naive proposal: 

Change `contract-out` so that the definition of its wrapper function appears in fully-expanded code as, well, the definition of the wrapper function. `(define-values (some-id) (lambda ___))`.

That way `syncheck:add-definition-target` would not need to change, IIUC, it would start finding these like any other function?

**Comment by samth** [comment]
Here's what I did to test:

1. Create a new file in DrRacket, with the contents
```
#lang racket
(require rnrs/base-6)
div0
```
2. Right click on `div0`.
3. Choose "Open Defining File"
4. Wait until the green dot appears in the other file.
5. Go back to the first tab.
6. Right click on `div0` again, see "Jump to Definition (in other file)".

**Comment by rfindler** [comment]
Okay! Thanks to some help from @mflatt I think this is fixed by a change to check syntax and a change to the contract system. I'll push shortly.

**Comment by greghendershott** [comment]
@rfindler @mflatt Thanks!

I'll try the next snapshot build that contains those commits.

**Comment by greghendershott** [comment]
Just tried it on snapshot `20210318-56970fd`. Works great! Thanks!!!

**Comment by rfindler** [comment]
Okay, this is totally embarrassing to me! You have been asking for this for ..... how many years?!?!?!?!? I just never understood that it was literally a bug!

**Comment by greghendershott** [comment]
Until recently I'm not sure I understood the pieces well enough to ask for this specifically.

What's extremely embarrassing to _me_ is that back in the day I learned Dr Racket only had "Open Defining File". And then in more recent years, somehow missed that it added "Jump to Definition (in other file)".

By the time I started using `syncheck`, `add-definition-target` had already been added to it, and somehow I never put 2+2 together to realize _why_ it had been added. (Some git history viewing yesterday made me go, "Oh!")

So. Someday, when it is safe to enjoy a beverage together without killing each other, I propose we have a mutual embarrassment toast. :smile: :champagne:

**Comment by rfindler** [comment]
You are too kind. :)

I am in!

Thanks, Greg.

**Comment by rfindler** [comment]
PS: the fact that jumping to definitions is so well hidden is bad UI
design, I call dibs on embarrassment for that one! ;)
